third_party/amd/lib/TritonAMDGPUTransforms/PointerCanonicalizer.cpp

```diff
@@ -0,0 +1,506 @@
+      Block *trueDest = condBrOp.getTrueDest();
+      // Walk the destination block only if you don't have visited it yet
+      if (isFalseOperand) {
+        falseOperands.push_back(basePtr);
+        falseOperands.push_back(offset);
+        Value falseDestArg = falseDest->getArgument(
+            operandNum - condBrOp.getNumTrueOperands() - 1);
+        if (!pointers.contains(falseDestArg)) {
+          nextPtr = falseDestArg;
+          Value basePtrArg = falseDest->addArgument(basePtr.getType(), curLoc);
+          Value offsetArg = falseDest->addArgument(offset.getType(), curLoc);
+          pointers[nextPtr] = FatPtr{basePtrArg, offsetArg, fatPtr.canNarrow};
+        }
+      } else {
+        trueOperands.push_back(basePtr);
+        trueOperands.push_back(offset);
+        Value trueDestArg = trueDest->getArgument(operandNum - 1);
+        if (!pointers.contains(trueDestArg)) {
+          nextPtr = trueDestArg;
+          Value basePtrArg = trueDest->addArgument(basePtr.getType(), curLoc);
+          Value offsetArg = trueDest->addArgument(offset.getType(), curLoc);
+          pointers[nextPtr] = FatPtr{basePtrArg, offsetArg, fatPtr.canNarrow};
+        }
+      }
+
+      // Create a new condBranch. We cannot simply extend the operands,
+      // because this would invalidate other operands pointing at the same
+      // cond branch
+      Value tensorPtr = createTensorPointer(fatPtr, curLoc);
+      auto newCondBranch = rewriter.create<cf::CondBranchOp>(
+          curLoc, condBrOp.getCondition(), trueDest, trueOperands, falseDest,
+          falseOperands);
+
+      newCondBranch.setOperand(operandNum, tensorPtr);
+      rewriteOpMap[condBrOp] = newCondBranch;
+      opToDelete.insert(condBrOp);
+    } else if (auto branchOp = dyn_cast<cf::BranchOp>(curOp)) {
+      size_t operandNum = curOperand->getOperandNumber();
+
+      FatPtr fatPtr = pointers[curOperand->get()];
+      Value offset = fatPtr.offset;
+      Value basePtr = fatPtr.basePtr;
+
+      branchOp.getDestOperandsMutable().append({basePtr, fatPtr.offset});
+      Value tensorPtr = createTensorPointer(fatPtr, curLoc);
+      branchOp->setOperand(operandNum, tensorPtr);
+      Block *dest = branchOp.getDest();
+
+      // Walk the destination block only if you don't have visited it yet
+      if (!pointers.contains(dest->getArgument(operandNum))) {
+        Value basePtrArg = dest->addArgument(basePtr.getType(), curLoc);
+        Value offsetArg = dest->addArgument(offset.getType(), curLoc);
+        nextPtr = dest->getArgument(operandNum);
+        pointers[nextPtr] = {basePtrArg, offsetArg, fatPtr.canNarrow};
+      }
+    } else if (isa<triton::PtrToIntOp>(curOp)) {
+      // Do nothing, just keep going
+    } else {
+      // If we meet an unsupported operation, materialize the fat pointer and
+      // continue. We flag that we met unknown usage of the pointer, hence we
+      // cannot ensure that all the pointers have been canonicalized
+      unknownOp = true;
```

Comment:
Seems no one is using `unknownOp`. But materialize should be valid. Still we should have some diagnostic that we may be missing an optimization here. Perhaps a DEBUG only assert.

Reply:
Well this is a diagnostic that we might use when trying to use buffer_loads. I.e., if we failed to remove all the uses of the pointer, we cannot issue a buffer_load. If you want, I can remove this now and add it back when we need it

Reply:
Not necessary, the pass is valid even if we have to materialize. 
I think it would be good to catch cases we missed though, thus an assert only in Debug. For example, if a new op is added.

Reply:
I can add the debug assert for now. Then when we add the buffer op logic, we can decide how to signal that the canonicalization didn't go well.